edk2toollib/uefi/edk2/build_objects/dsc.py

```diff
@@ -0,0 +1,455 @@
+        self.inf = inf
+        self.source_info = source_info
+
+    def __eq__(self, other):
+        if (type(other) is not library_class):
+            return False
+        # if they're both null
+        if self.libraryclass.lower() == "null" and other.libraryclass.lower() == "null":
+            return self.inf == other.inf
+        return self.libraryclass.lower() == other.libraryclass.lower()
+
+    def __hash__(self):
+        # if we're a null lib, we want the hash to be based on the inf path
+        if (self.libraryclass.lower() == "null"):
+            return hash(self.inf)
+        else:
+            return hash(self.libraryclass)
+
+    def __repr__(self):
+        return f"{self.libraryclass}|{self.inf} @ {self.source_info}"
+
+
+class pcd:
+    ''' Contains the data for a specific pcd '''
+    ''' PcdTokenSpaceGuidCName.PcdCName|Value '''
+
+    def __init__(self, namespace, name, value, source_info=None):
+        self.namespace = namespace
+        self.name = name
+        self.value = value
+        self.source_info = source_info
+
+    def __eq__(self, other):
+        if not issubclass(other.__class__, pcd):
+            return False
+        return self.namespace == other.namespace and self.name == other.name
+
+    def __hash__(self):
+        return hash(f"{self.namespace}.{self.name}")
+
+    def __repr__(self):
+        return f"{self.namespace}.{self.name} = {self.value} @ {self.source_info}"
+
+
+class pcd_typed(pcd):
+    ''' PcdTokenSpaceGuidCName.PcdCName|Value[|DatumType[|MaximumDatumSize]] '''
+
+    def __init__(self, namespace, name, value, datum_type, max_size=0, source_info=None):
+        super().__init__(namespace, name, value, source_info)
+        self.datum_type = datum_type
+        self.max_size = int(max_size)
+
+    def __repr__(self):
+        return f"{self.namespace}.{self.name} = {self.value} |{self.datum_type}|{self.max_size} @ {self.source_info}"
+
+
+pcd_variable_attributes = ["NV", "BS", "RT", "RO"]
+
+
+class pcd_variable(pcd):
+    ''' PcdTokenSpaceGuidCName.PcdCName|VariableName|VariableGuid|VariableOffset[|HiiDefaultValue[|HiiAttrubte]] '''
+
+    def __init__(self, namespace, name, var_name, var_guid, var_offset, default=None, attributes=[], source_info=None):
+        super().__init__(namespace, name, "", source_info)
+        self.var_name = var_name
+        self.var_guid = var_guid
+        self.var_offset = var_offset
+        self.default = default
+
+        if type(attributes) is str:
+            attributes = attributes.split(",")
+        attributes = [str(x).upper().strip() for x in attributes]
+        if any([x not in pcd_variable_attributes for x in attributes]):
+            raise ValueError(f"Invalid PcdHiiAttribute values: {attributes}")
+        self.attributes = attributes
+
+    def __repr__(self):
+        return f"{self.namespace}.{self.name} = {self.var_name} |{self.var_guid}|{self.var_offset}|{self.default}|{self.attributes} @ {self.source_info}"
+
```

Comment:
should there be a class for structured pcd to store pcd info and pcd fields' value?

Reply:
That makes sense, but I hadn't tackled that yet. It there a document that describes the format for structured PCD's?

Reply:
I'll add this as an issue. #83